### DISCUSSION TITLE
Add function providing product shipping options

### DISCUSSION THREAD

**Discussion by Hlavtox** [PrestaShop/PrestaShop#33420] [Ideas]
### Prerequisites

- [X] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md)
- [X] I have already [searched in existing features request](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3AImprovement%2CFeature%2CDeveloper+Feature+) and found no previous suggestion of this feature.

### Is your feature request related to a problem?

There are many cases in modules, where we need to get available shipping options for a single product (carrier and price).

There is no such function that provides that. There is such a function for cart, but not for a single product.

Usage examples:
- If you want to display shipping options on product page, for example on another product tab. Many shops use this.
- If you are exporting products to XML feeds and you want delivery options to be there, so you dont have to set them manually in merchant center.
- If you want to show a custom free shipping flag on product.

### Describe the solution you'd like

Add a new function to Carrier class.

**Parameters**
  - id_product or Product object
  - id_zone - so we can use the default country's zone or use customer's, if he has his address set up.

**Return**
It would return an array with available shipping options. Each option would contain Carrier object and the calculated price of shipping for this single product.

### Alternatives you've considered

To build this function over and over again in every module.

### Additional context

Ping @kpodemski, maybe you could provide some more examples of use?

**Comment by hibatallahAouadni** [comment]
Hello @Hlavtox 

Thank you for your suggestion. The Developer Team will take it into consideration for future developments. 
Please be aware that there is no guarantee that this feature will be developed anytime soon.

If you are willing to implement this feature into PrestaShop as a [Pull Request](https://devdocs.prestashop.com/1.7/contribute/contribute-pull-requests/), the teams will be happy to help you by ensuring that it is appropriate and coherent with the rest of the project. If that is the case, please let us know! 

Thank you

**Comment by kpodemski** [comment]
Most of the time, it works using the entire Cart because Cart contents directly impact shipping price. I agree that it could be helpful for external feeds and simple calculations when you don't need to provide an address and other details.

From my experience, I know these calculations are very time-consuming, and having a reliable method that is light and more performant could be helpful.

**Comment by matks** [comment]
> Most of the time, it works using the entire Cart because Cart contents directly impact shipping price. 

Does the cart content also impact the options, like exclude some?

**Comment by Hlavtox** [comment]
@matks Not directly, there is nothing in the cart itself that would directly affect anything. 👍

**Comment by Hlavtox** [comment]
Basically we just want to simulate “what shipping options and prices do I get, when I add this single item in cart”

**Comment by Hlavtox** [comment]
**Example 1 - scroll down to delivery and returns. Actually lot of my customers asked me to do this.**
https://www.vagabond.com/fr/quincy-5285-050-07/

**Example 2 - product shipping prices in Google and Facebook XML feeds**
https://support.google.com/merchants/answer/6324484?hl=fr&ref_topic=6324338
